src/hotspot/cpu/riscv/stubGenerator_riscv.cpp

```diff
@@ -5042,6 +5042,391 @@ class StubGenerator: public StubCodeGenerator {
+      __ vwredsumu_vs(vtemp2, vs2acc[2], vzero);
+
+    // Extracting results for:
+    // s1_new
+    __ vmv_x_s(temp0, vs1acc[0]);
+    __ add(s1, s1, temp0);
+    // s2_new
+    __ vsetvli(temp0, count, Assembler::e32, Assembler::m1);
+    __ vmv_x_s(temp1, vtemp1);
+    __ add(s2, s2, temp1);
+    if (MaxVectorSize == 16) {
+      __ vmv_x_s(temp2, vtemp2);
+      __ add(s2, s2, temp2);
+    }
+  }
+
+  void adler32_process_bytes_by32(Register buff, Register s1, Register s2, Register count,
+    VectorRegister vtable, VectorRegister vzero, VectorRegister *vbytes, VectorRegister *vs1acc, VectorRegister *vs2acc, 
+    Register temp0, Register temp1, Register temp2, VectorRegister vtemp1, VectorRegister vtemp2) {
+
+    // Same algorithm as for 64 bytes, but shrinked to 32 bytes step:
+    //   s1_new = s1 + b1 + b2 + ... + b32
+    //   s2_new = s2 + (s1 + b1) + (s1 + b1 + b2) + ... + (s1 + b1 + b2 + ... + b32)
+    //          = s2 + s1 * 32 + (b1 * 32 + b2 * 31 + ... + b32 * 1)
+    //          = s2 + s1 * 32 + (b1, b2, ... b16) dot (32, 31, ... 1)
+
+    __ mv(count, 32);
+    // Load data
+    __ vsetvli(temp0, count, Assembler::e8, Assembler::m2);
+    __ vle8_v(vbytes[0], buff);
+    __ addi(buff, buff, 32);
+
+    // Reduction sum for s1_new
+    // 0xFF * 32 = 0x1FE0, so:
+    // 1. Need to do vector-widening reduction sum
+    // 2. It is safe to perform sign-extension during vmv.x.s with 16-bits elements
+    __ vwredsumu_vs(vs1acc[0], vbytes[0], vzero);
+    // Multiplication for s2_new
+    __ vwmulu_vv(vs2acc[0], vtable, vbytes[0]);
+
+    // s2 = s2 + s1 * 32
+    __ slli(temp1, s1, 5);
+    __ add(s2, s2, temp1);
+
+    // Summing up calculated results for s2_new
+    __ vsetvli(temp0, count, Assembler::e16, Assembler::m2);
+    // Upper bound for reduction sum:
+    // 0xFF * (32 + 31 + ... + 2 + 1) = 0x20DF0 max for whole register group, so:
+    // 1. Need to do vector-widening reduction sum
+    // 2. It is safe to perform sign-extension during vmv.x.s with 32-bits elements
+    __ vwredsumu_vs(vtemp1, vs2acc[0], vzero);
+    if (MaxVectorSize == 16)
+      // For small vector length, the rest of multiplied data
+      // is in successor of vs2acc[i], so summing it up, too
+      __ vwredsumu_vs(vtemp2, vs2acc[1], vzero);
+
+    // Extracting results for:
+    // s1_new
+    __ vmv_x_s(temp0, vs1acc[0]);
+    __ add(s1, s1, temp0);
+    // s2_new
+    __ vsetvli(temp0, count, Assembler::e32, Assembler::m1);
+    __ vmv_x_s(temp1, vtemp1);
+    __ add(s2, s2, temp1);
+    if (MaxVectorSize == 16) {
+      __ vmv_x_s(temp2, vtemp2);
+      __ add(s2, s2, temp2);
+    }
+  }
+
+  void adler32_process_bytes_by16(Register buff, Register s1, Register s2, Register right_16_bits,
+    VectorRegister vtable, VectorRegister vzero, VectorRegister *vbytes, VectorRegister *vs1acc, VectorRegister *vs2acc, 
+    Register temp0, Register temp1, Register temp2, VectorRegister vtemp1, VectorRegister vtemp2, int LMUL) {
```

Comment:
Let's remove this `LMUL` param as all the callsites now passes value 1. 

Did you considered unifying adler32_process_bytes_by16/32/64 into one function with one extra param indicating the size? Seems to me that they duplicate most of the code. I guess it should be no big difference for the 16 variant to do vector-widening reduction sum at the end, just like the other two?

BTW: I can help test the performance difference as I have just added Banana-PI into my RV testing army.

Reply:
Thanks, I unified the code for all steps sizes. Can you please do a re-run on Banana PI for performance?